### ISSUE TITLE
Profile: model and API

### ISSUE THREAD

**Issue by pdpinch** [mitodl/open-discussions#674]
As a user, I need a model for storing my profile information and a basic DRF API for updating it from the front-end. 

### Initial fields:
- [x] (full) name
- [ ] headline
- [ ] bio
- [ ] agreed to terms and conditions (not the same as `email_optin` ?)

### related issues
- #683 profile image (separate issue)
- MIT status (?) 

We expect to add additional profile fields later, when we have features that can use them.

**Comment by pdpinch** [comment]
Looking at the existing Profile model, it looks like we have a profile image already, but we'll have to add the other fields:

https://github.com/mitodl/open-discussions/blob/master/profiles/models.py#L15-L27

`first_name` and `last_name` are on the django User model, but aren't being populated currently and will need to be added to the API.

**Comment by mbertrand** [comment]
@pdpinch: 

- What status values are possible for `MIT status`?
- does it make sense to split the value of the `name` field into `first` and `last` name values during migration for existing profiles?
- Is `profile image` the same as the current `image` (and `image_small`, `image_medium`) field?

**Comment by pdpinch** [comment]
I think `MIT Status` is a boolean, so that we have the ability to distinguish MIT users in the UI. However, maybe we should get an opinion from @rhysyngsun since MIT users are also distinguished by their use of Touchstone authentication. 

We decided today that we're only going to do full names, which is already present on the profile. So, no change necessary. I'll update the issue description to match. 

For profile image, we need to be able to support users uploading new images. I don't think we can do that with the current model. I think we need something more like what we have in MicroMasters, where there's a foreign key to a ImageField model. Please take a look at https://github.com/mitodl/micromasters/blob/master/profiles/models.py#L161-L163 and then propose an approach here.

**Comment by pdpinch** [comment]
We're also considering pushing/pulling bios from MicroMasters, and using gravatar as an option for profile images. Stay tuned.

**Comment by rhysyngsun** [comment]
@mbertrand @pdpinch the `image*` fields on `Profile` are the absolute urls to the images in MM, which in MM are `models.ImageField`, which in the db is just a `VARCHAR`:

```

postgres=# \d profiles_profile
                                             Table "public.profiles_profile"
            Column             |           Type           |                           Modifiers                           
-------------------------------+--------------------------+---------------------------------------------------------------
 id                            | integer                  | not null default nextval('profiles_profile_id_seq'::regclass)
 account_privacy               | text                     | not null
 email_optin                   | boolean                  | not null
 edx_employer                  | character varying(255)   | 
 edx_job_title                 | character varying(255)   | 
 state_or_territory            | character varying(255)   | 
 edx_name                      | text                     | 
 edx_bio                       | text                     | 
 country                       | text                     | 
 edx_requires_parental_consent | boolean                  | 
 edx_level_of_education        | text                     | 
 edx_goals                     | text                     | 
 edx_language_proficiencies    | jsonb                    | 
 gender                        | character varying(6)     | 
 edx_mailing_address           | text                     | 
 user_id                       | integer                  | not null
 date_joined_micromasters      | timestamp with time zone | 
 date_of_birth                 | date                     | 
 filled_out                    | boolean                  | not null
 first_name                    | text                     | 
 last_name                     | text                     | 
 preferred_language            | text                     | 
 preferred_name                | text                     | 
 birth_country                 | text                     | 
 city                          | text                     | 
 student_id                    | integer                  | 
 agreed_to_terms_of_service    | boolean                  | not null
 verified_micromaster_user     | boolean                  | not null
 nationality                   | text                     | 
 image                         | character varying(100)   | 
 about_me                      | text                     | 
 phone_number                  | character varying(50)    | 
 postal_code                   | character varying(16)    | 
 romanized_first_name          | character varying(30)    | 
 romanized_last_name           | character varying(50)    | 
 updated_on                    | timestamp with time zone | 
 image_small                   | character varying(100)   | 
 image_medium                  | character varying(100)   | 
 address                       | character varying(100)   | 
 mail_id                       | uuid                     | not null
 fake_user                     | boolean                  | not null
Indexes:
    "profiles_profile_pkey" PRIMARY KEY, btree (id)
    "profiles_profile_mail_id_0e5c24ac_uniq" UNIQUE CONSTRAINT, btree (mail_id)
    "profiles_profile_student_id_key" UNIQUE CONSTRAINT, btree (student_id)
    "profiles_profile_user_id_a3e81f91_uniq" UNIQUE CONSTRAINT, btree (user_id)
Foreign-key constraints:
```

If we change these fields to also be `ImageField`s, migrate the values to be relative paths, and configure them identically to MM (with a S3 `upload_to`) I would expect them to work.

Regarding `MIT Status`, this sounds like a read-only field? We could have the serializer reflect that and return `True|False` depending on the existence of a touchstone social auth record. However, we should keep in mind that this would mean "this user at some point in the past or present was/is a MIT user", **not** "this user currently is a MIT user". I think that's an important distinction, but I'm not sure if it's material to the use of that value for this feature.

**Comment by mbertrand** [comment]
@rhysyngsun @pdpinch would it make sense to make `mit_status=True` if the user's email ends with `mit.edu`, otherwise query the moira service for the user's email to see if any record exists?

**Comment by rhysyngsun** [comment]
@mbertrand not sure on the former approach, but for the latter we'd want to cache that in the db, because I would expect that to be an expensive lookup that we wouldn't want to incur when rendering api responses. Also, are you going to be able to query Moira for users that were created from MM and won't have a touchstone auth (I'm not sure how those services relate, if at all)?

**Comment by mbertrand** [comment]
@rhysyngsun @pdpinch actually moira won't be of much use here, someone could be affiliated with MIT but not be on any lists.  Not sure if Touchstone is that helpful either, anybody can create a Touchstone collaboration account.

**Comment by pdpinch** [comment]
Good point about Touchstone collaborator accounts. And I agree that moira isn't helpful. 

Let's remove the MIT status from this issue to give us some more time to determine a) whether it's necessary and b) how to implement it. I'll update the description.

**Comment by pdpinch** [comment]
I'm also going to move the profile image into a separate issue, because I think it needs more attention.

**Comment by mbertrand** [comment]
@pdpinch @rhysyngsun should the API also allow for viewing a list of profiles, or just individual profiles?  (The API already can show a list of users including profiles, for those with appropriate permissions).

**Comment by rhysyngsun** [comment]
@mbertrand one thing to consider is security around some fields if other users are going to be able to query this API. Nothing super-private here yet I don't think, but we maybe wouldn't want to show TOS status to other users. We have a way of handling this over in MM where we swap out different profile serializers depending on the user requesting it so maybe we'd want to port that mechanism to here?

**Comment by mbertrand** [comment]
@rhysyngsun Currently the API in the PR only allows users to view their own profile (unless staff or superuser). A list of profiles isn't available except indirectly via the user API (which requires jwt-staff permissions).  I also set the TOS optin field to write-only in the serializer, to match the email optin field, so it won't appear even to that user, but that's an assumption I made while working on the PR and forgot to ask about, so let me know if that should change.
